epinar/assignment2/ancestors.py

```diff
@@ -0,0 +1,113 @@
+class TreeNode:
+	def __init__(self, data):
+		self.data = data
+		self.left = None
+		self.right = None
+
+# Initiates a tree with an input value. Keeps the nodes in a dictionary for fast reach.
+# Each data(key) value should be unique.
+class BinaryTree:
+	def __init__(self, data):
+		self.tree = {}
+		node = TreeNode(data)
+		self.head = node
+		self.tree[data] = node
+		self.a =[] # shared variable between classes
+	# We assume that parent node exists and the right/left child is free.
+	def add_left(self, parent, data):
+		p = self.tree[parent]
+		node = TreeNode(data)
+		p.left = node
+		self.tree[data] = node
+	def add_right(self, parent, data):
+		p = self.tree[parent]
+		node = TreeNode(data)
+		p.right = node
+		self.tree[data] = node
+
+	# Prints in dictionary order.
+	def print_dict(self):
+		for k, v in self.tree.items():
+			print("node: ", v.data)
+			if v.right:
+				print(" r: ", v.right.data)
+			if v.left:
+				print(" l: ", v.left.data)
+
+	# Prints the tree in order.
+	def print(self):
+		self._print(self.head)
+
+	def _print(self, curr):
+		if not curr:
+			return
+		print(curr.data)
+		self._print(curr.left)
+		self._print(curr.right)
+
+	# Returns the ancestors ordered from lowest to highest in tree.
+	def ancestors(self, data):
+		self.a = []
```

Comment:
Mmh, I do not understand the role of this a. The name is confusing, please avoid one-letter names except for iterators and counters:
https://google.github.io/styleguide/pyguide.html?showone=Naming#Naming

Here I think you attaching to the tree data whose lifetime is the run of the ancestors() method. I do not think you should be attaching the state of the ancestors computation to the tree itself. For instance, this make the ancestors() method incompatible with concurrent programming where, for example, 2 threads would be stepping on each other toes by both appending to the same self.a variable.

I think you are solving the problem of sharing data between the runs of a recursive function. In this case, a is called in general, an accumulator. Try to be explicit instead of implicit: here, pass the accumulator via the function arguments, instead of passing some arguments with the function calls, and others via the root object (this solution has the benefit to solve the problem of the concurrency).

Reply:
Here, instead of printing the nodes, I created this variable to accumulate them. Then, I learnt a nice way to do that in python: by yielding. I will try to apply that wherever I see needed next times.